### PR TITLE
fix(metadata): align metadata suffix hash between turbopack

### PR DESCRIPTION
**Pull request by kwonoj**
### What?

Wraps up metadata-dynamic-routes tests fixes for the turbopack. There is 1 remaining failing test due to lacks of supporting `import.meta.url` which need to be addressed separately.

I spent amount of time why turbopack cannot find the route for the dynamic metadata for a certain route. In the end, found there are mismatching expectations for the route due to different hash for the certain route. We do use the same djb2 hash between next.js and turbopack both, so it was quite confusing why we don't get deterministic hash.

After trying some experiments, found out root cause was how 2 different runtimes handle overflow for given type of numbers. In rust + turbpack we use u32 and do 32-bit hash calculation for given string, while in js we implicitly used number type as is, in result overflow occurs with default 53-bit float. 

Originally I tried to adjust hash in turbopack side to preserve js hash as-is, but so far I found it was non trivial to do so as rust there's no out of the box types we can coerce to the js number type. In result, unlike other fixes in turbopack this PR changes how js hash is being generated. I hope this woulndn't be a breaking changes; expect so since this is a metadata specific hash that we do not have written spec for it.

Closes WEB-1890

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ad23fe211cf371c05182d84448b50c76626e09c7

`pnpm test-start test/e2e/app-dir/metadata-dynamic-routes/index.test.ts` 
- app dir - metadata dynamic routes > should optimize routes without multiple generation API as static routes

<details>
<summary>Expand output</summary>

  ● app dir - metadata dynamic routes › should optimize routes without multiple generation API as static routes

    expect(received).toMatchObject(expected)

    - Expected  - 1
    + Received  + 0

      Object {
    -   "/(group)/dynamic/[size]/apple-icon-48jo90/[[...__metadata_id__]]/route": "app/(group)/dynamic/[size]/apple-icon-48jo90/[[...__metadata_id__]]/route.js",
        "/gsp/sitemap/[__metadata_id__]/route": "app/gsp/sitemap/[__metadata_id__]/route.js",
        "/sitemap.xml/route": "app/sitemap.xml/route.js",
        "/twitter-image/route": "app/twitter-image/route.js",
      }

      517 |         )
      518 |
    > 519 |         expect(appPathsManifest).toMatchObject({
          |                                  ^
      520 |           // static routes
      521 |           '/twitter-image/route': 'app/twitter-image/route.js',
      522 |           '/sitemap.xml/route': 'app/sitemap.xml/route.js',

      at Object.toMatchObject (e2e/app-dir/metadata-dynamic-routes/index.test.ts:519:34)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| buildDuration | 11s | 11.4s | ⚠️ +397ms |
| buildDurationCached | 6.6s | 6.1s | N/A |
| nodeModulesSize | 175 MB | 175 MB | ⚠️ +2.36 kB |
| nextStartRea..uration (ms) | 431ms | 425ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| 199-HASH.js gzip | 32.3 kB | 32.4 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.2 kB | 53.2 kB | ✓ |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.5 kB | 45.5 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 252 B | N/A |
| main-HASH.js gzip | 35.3 kB | 35.3 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 98.9 kB | 98.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.59 kB | 2.59 kB | ✓ |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.38 kB | 4.38 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 316 B | 318 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.99 kB | 3.99 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| index.html gzip | 528 B | 531 B | N/A |
| link.html gzip | 541 B | 543 B | N/A |
| withRouter.html gzip | 524 B | 526 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| edge-ssr.js gzip | 96 kB | 96 kB | N/A |
| page.js gzip | 140 kB | 140 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js metadata-dynamic | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 623 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 25.8 kB | 25.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>199-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: b0cf6b560a39fdf8e629adc4815ec279ea341aac

File: packages/next-swc/crates/next-core/src/next_app/metadata/mod.rs [hunk 1]
```diff
@@ -273,7 +273,8 @@ fn format_radix(mut x: u32, radix: u32) -> String {
         }
     }
 
-    result.into_iter().rev().collect()
+    result.reverse();
+    result[..6].iter().collect()
```
Comment (jridgewell) [inline packages/next-swc/crates/next-core/src/next_app/metadata/mod.rs:277]: Nit: I think you can
```suggestion
    result.into_iter().rev().take(6).collect()
```

**Review by jridgewell — APPROVED** [review]

**Comment by ForsakenHarmony** [comment]
was actually looking into this and the rust version that more closely matches the js version would have been

```rs
fn djb2_hash(str: &str) -> u64 {
    let mut hash: f64 = 5381.0;
    for char in str.chars() {
        let code = char as u64 as f64;
        let tmp_hash = (hash as i64 as i32) << 5;
        hash = (tmp_hash as f64) + hash + code; 
    }
    hash.abs() as u64
}
```

but changing the js version is probably better

File: packages/next-swc/crates/next-core/src/next_app/metadata/mod.rs (hunk 1 above)
Comment (ForsakenHarmony) [inline packages/next-swc/crates/next-core/src/next_app/metadata/mod.rs:277]: this should happen in `get_metadata_route_suffix` below, not in here

**Pull request merged**